### PR TITLE
[WIP] Adding support for docker stack

### PR DESCRIPTION
**Pull request by vidarl**
FYI : This branch is now based on https://github.com/ezsystems/ezplatform/pull/192

Todo:
- [ ] move to ezplatform repo _(and merge back up to other repos)_, code placed here during testing to be able to test with demo content and with ee repos

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -21,6 +21,17 @@ REDIS_IMAGE=redis
...
+DATASET_DBDUMP_IMAGE=my-ez-app-dataset-dbdump
+
+DATASET_VARDIRDUMP_DOCKER_FILE=Dockerfile.vardirdump
+DATASET_VARDIRDUMP_IMAGE=my-ez-app-dataset-vardirdump
+
+DATASET_VARDIR=my-ez-app
 
 # Install config, used by .platform.app.yaml among others
-INSTALL_EZ_INSTALL_TYPE=studio-clean
+INSTALL_EZ_INSTALL_TYPE=demo
```
Comment (andrerom) [inline .env]: this should be applied to master btw, as it's a mistake.

File: Dockerfile [hunk 2]
```diff
@@ -14,16 +11,18 @@ RUN if [ -d .git ]; then echo "ERROR: .dockerignore folders detected, exiting" &
 
 # Install and prepare install
 RUN mkdir -p web/var \
-    && composer install --optimize-autoloader --no-progress --no-interaction --no-suggest --prefer-dist \
+    # For now, only run composer in order to generate parameters.yml
+    && composer run-script build --no-interaction \
```
Comment (andrerom) [inline Dockerfile]: should then add `composer dump -o` as well then to make sure autoload is optimized.

File: doc/docker-compose/import-dataset.yml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    command: >
+      /bin/bash -c "
+      time mysql -u $${DATABASE_USER} -p$${DATABASE_PASSWORD} $${DATABASE_NAME} -h db< /dbdump/ezp.sql"
+
+  import-vardir:
+    image: ${DATASET_VARDIRDUMP_IMAGE}
+    volumes:
+     - vardir:/nfs/vardirs
+    environment:
+     - DATASET_VARDIR
```
Comment (andrerom) [inline doc/docker-compose/import-dataset.yml]: should this rather be `DATASET_VARDIR =${APP_PROD_IMAGE}` to avoid duplicating the variable across two env variables?

File: doc/docker-compose/install.yml [hunk 4]
```diff
@@ -28,16 +28,19 @@ services:
...
     # Second chown line:  For dev and behat tests we give a bit extra rights, never do this for prod.
+    # assetic:dump and assets:install is run with using hard copy instead of symlinks. Needed by prod images
     command: >
       /bin/sh -c "
       composer install --no-progress --no-interaction --prefer-dist --no-suggest --optimize-autoloader;
       mkdir -p web/var;
       php /scripts/wait_for_db.php;
       php app/console ezplatform:install ${INSTALL_EZ_INSTALL_TYPE};
+      php app/console assetic:dump --env=prod;
+      php app/console assets:install --env=prod;
```
Comment (andrerom) [inline doc/docker-compose/install.yml]: you can most likly skip these two and rather run this before composer install:
`composer config extra.symfony-assets-install hard`

But you'll then need to checkout the original file again at the end to not change local setup, in multistage build we would not have such problems.

File: .env (hunk 1 above)
Comment (vidarl) [inline .env]: Moved to separate PR https://github.com/ezsystems/ezplatform-ee-demo/pull/35

Will remove comitt from this branch once PR #35 is merged to master

File: Dockerfile (hunk 2 above)
Comment (vidarl) [inline Dockerfile]: Added line as suggested

File: doc/docker-compose/demo.yml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+version: '2'
```
Comment (andrerom) [inline doc/docker-compose/demo.yml]: needs to be same version as the others

File: Dockerfile.nginx [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+ENV SYMFONY_ENV=prod
+
+# Copy in project files into work dir
+COPY . /var/www
+
+# assetic:dump requires parameters.yml, so let's generate it
+RUN composer run-script build --no-interaction
+
+# @AR : I didn't get this working, so reverting to asset:install
+# RUN composer config extra.symfony-assets-install hard
```
Comment (andrerom) [inline Dockerfile.nginx]: I assume you did that before `composer run-script`?

File: Dockerfile.nginx [hunk 7]
```diff
@@ -0,0 +1,31 @@
+
+FROM ezsystems/php:7.1-v1 as app
+
+ENV SYMFONY_ENV=prod
+
+# Copy in project files into work dir
+COPY . /var/www
+
+# assetic:dump requires parameters.yml, so let's generate it
+RUN composer run-script build --no-interaction
```
Comment (andrerom) [inline Dockerfile.nginx]: `build` => `post-install-cmd`, safer for future changes. Symfony Flex uses `symfony-scripts`, goes for other places as well

File: .dockerignore [hunk 8]
```diff
@@ -1,6 +1,9 @@
 ### IDE files
 .idea/
 
+### Composer files
+auth.json
```
Comment (andrerom) [inline .dockerignore]: I assume this breaks install.yml?

Comment (andrerom) [inline .dockerignore]: it will also break dev usage of composer _(with dev config mounting folder, afaik mount also takes entries here into account)_

With multi stage we can safely remove it in build stage anyway right?

**Comment by kamilmusial** [comment]
Are some of those changes related to `ezplatform-ee` repo as well? If yes, shouldn't some of them be made on clean repo and then merged up with the rest here?

**Comment by andrerom** [comment]
@kamilmusial Updated description, @vidarl is on vacation until next week.

File: .dockerignore (hunk 8 above)
Comment (vidarl) [inline .dockerignore]: nope. This does not break anything. In install.yml and in dev usage we bind mount the project directly to the container. So we still got access to auth.json there.

File: Dockerfile.nginx (hunk 7 above)
Comment (vidarl) [inline Dockerfile.nginx]: You have now added a doc/docker/Dockerfile-nginx file.
That obsoletes this file entirely and I'll remove it

**Comment by vidarl** [comment]
@andrerom : I feel pretty done with this now. PR is based on https://github.com/ezsystems/ezplatform/pull/192. Not sure it is worth doing a rebase (to fix conflicts etc) before that one is merged to ezplatform and down to the ezplatform-ee-demo meta repo?
